feat(controller): Support GC for memoization caches

terrytangyuan opened the pull request:
Fixes https://github.com/argoproj/argo-workflows/issues/5406. Originally part of https://github.com/argoproj/argo-workflows/pull/5537 and on top of the previous work on ConfigMaps informer.

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>

NikeNano commented on workflow/controller/controller.go:
```diff
@@ -801,6 +809,75 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+					log.Error("Unable to convert object to configmap when syncing ConfigMaps")
+					continue
+				}
+				err := wfc.cleanupUnusedCache(cm, gcAfterNotHitDuration)
+				if err != nil {
+					wfc.eventRecorderManager.Get(cm.GetNamespace()).Event(cm, apiv1.EventTypeWarning, "SyncFailed", err.Error())
+					log.WithError(err).Errorf("Unable to sync ConfigMap: %s", cm.Name)
+					continue
```
```suggestion
				if err := wfc.cleanupUnusedCache(cm, gcAfterNotHitDuration); err != nil {
					wfc.eventRecorderManager.Get(cm.GetNamespace()).Event(cm, apiv1.EventTypeWarning, "SyncFailed", err.Error())
					log.WithError(err).Errorf("Unable to sync ConfigMap: %s", cm.Name)
					continue
```

NikeNano commented on workflow/controller/controller.go:
```diff
@@ -311,6 +311,14 @@ func (wfc *WorkflowController) startLeading(ctx context.Context, logCtx *log.Ent
 	}
+	if v, found := os.LookupEnv("CACHE_GC_PERIOD"); found {
+		cacheGCPeriod, err := time.ParseDuration(v)
+		if err != nil {
+			log.WithField("CACHE_GC_PERIOD", v).WithError(err).Panic("failed to parse")
+		} else {
+			go wait.UntilWithContext(ctx, wfc.syncAllCacheForGC, cacheGCPeriod)
+		}
```
maybe we could remove the else statement? 

```suggestion
		if err != nil {
			log.WithField("CACHE_GC_PERIOD", v).WithError(err).Panic("failed to parse")
			return 
		} 
		go wait.UntilWithContext(ctx, wfc.syncAllCacheForGC, cacheGCPeriod)
```

NikeNano commented on workflow/controller/controller.go:
```diff
@@ -801,6 +809,75 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+	var modified bool
+	for key, rawEntry := range cm.Data {
+		var entry controllercache.Entry
+		err := json.Unmarshal([]byte(rawEntry), &entry)
+		if err != nil {
+			return fmt.Errorf("malformed cache entry: could not unmarshal JSON; unable to parse: %w", err)
+		}
```
```suggestion
		if err := json.Unmarshal([]byte(rawEntry), &entry); err != nil {
			return fmt.Errorf("malformed cache entry: could not unmarshal JSON; unable to parse: %w", err)
		}
```

sarabala1979 commented:
@terrytangyuan Mainly this feature is trying to solve the configmap(size). Do we need to add this complexity to the workflow controller?  you can create a separate configmap for each cache to avoid the growing size. only thing configmap can have owenerreference of workflow if user want TTL/GC for cache. configmap will delete once workflow deleted.

CacheGC and Maxage will be out of sync in this feature.

alexec reviewed (changes requested):
changes requested

terrytangyuan commented on workflow/controller/controller.go:
```diff
@@ -801,6 +809,73 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+		cm.Namespace, cm.Name)
+	if len(cm.Data) == 0 {
+		log.Infof("Deleting ConfigMap %s since it doesn't contain any cache entries", cm.Name)
+		request := wfc.kubeclientset.CoreV1().RESTClient().Delete().RequestURI(selfLink)
```
Updated.

terrytangyuan commented on workflow/controller/controller.go:
```diff
@@ -801,6 +809,73 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+				}
+				if err := wfc.cleanupUnusedCache(cm, gcAfterNotHitDuration); err != nil {
+					wfc.eventRecorderManager.Get(cm.GetNamespace()).Event(cm, apiv1.EventTypeWarning, "SyncFailed", err.Error())
+					log.WithError(err).Errorf("Unable to sync ConfigMap: %s", cm.Name)
```
Updated.

terrytangyuan commented on workflow/controller/controller.go:
```diff
@@ -801,6 +809,73 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+					continue
+				}
+				if err := wfc.cleanupUnusedCache(cm, gcAfterNotHitDuration); err != nil {
+					wfc.eventRecorderManager.Get(cm.GetNamespace()).Event(cm, apiv1.EventTypeWarning, "SyncFailed", err.Error())
```
Removed

terrytangyuan commented on workflow/controller/controller.go:
```diff
@@ -801,6 +809,73 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+			log.WithField("CACHE_GC_AFTER_NOT_HIT_DURATION", v).WithError(err).Panic("failed to parse")
+		} else {
+			log.Info("Cache GC is enabled. Syncing all cache for GC.")
+			configMaps := wfc.configMapInformer.GetIndexer().List()
```
This `configMapInformer` only contains the configmaps with the correct label keys so `List()` is equivalent.

terrytangyuan commented on workflow/controller/controller.go:
```diff
@@ -801,6 +809,73 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+	} else {
+		if modified {
+			log.Infof("Modified ConfigMap: %s: %v", cm.Name, cm)
+			request := wfc.kubeclientset.CoreV1().RESTClient().Put().RequestURI(selfLink).Body(cm)
```
Updated

alexec commented on docs/environment-variables.md:
```diff
@@ -13,6 +13,8 @@ Note that these environment variables may be removed at any time.
 | `ARCHIVED_WORKFLOW_GC_PERIOD` | `time.Duration` | `24h` | The periodicity for GC of archived workflows. |
 | `ARGO_PPROF` | `bool` | `false` | Enable pprof endpoints |
 | `ARGO_TRACE` | `string` | `"1"` | Whether to enable tracing statements in Argo components. |
+| `CACHE_GC_PERIOD` | `time.Duration` | | How often to perform memoization cache GC. |
```
default is false than?

alexec commented on workflow/controller/controller.go:
```diff
@@ -798,6 +806,65 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
 }
 
+func (wfc *WorkflowController) syncAllCacheForGC(ctx context.Context) {
+	if v, found := os.LookupEnv("CACHE_GC_AFTER_NOT_HIT_DURATION"); found {
```
use `envutil.GetDuration`?

alexec commented on workflow/controller/cache/configmap_cache.go:
```diff
@@ -117,6 +118,7 @@ func (c *configMapCache) Save(ctx context.Context, key string, nodeId string, va
 	}
 
 	creationTime := time.Now()
+	cache.SetLabels(map[string]string{indexes.LabelKeyConfigMapType: indexes.LabelValueCacheTypeConfigMap})
```
I've another purpose for this informer (plugins), so it should contain all config maps with the label, I think you'll need to refactor?

terrytangyuan replied:
This PR only addes GC for configmap cache. If we want to support GC for other types of configmaps, it might be better to add later on.

alexec commented on workflow/controller/controller.go:
```diff
@@ -798,6 +806,65 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+	}
+}
+
+func (wfc *WorkflowController) cleanupUnusedCache(ctx context.Context, cm *apiv1.ConfigMap, gcAfterNotHitDuration time.Duration) error {
```
put in own file

alexec commented on workflow/controller/controller.go:
```diff
@@ -798,6 +806,65 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
 	return true
 }
 
+func (wfc *WorkflowController) syncAllCacheForGC(ctx context.Context) {
```
put in own file

alexec commented on workflow/controller/controller.go:
```diff
@@ -798,6 +806,65 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+		}
+	}
+	if len(cm.Data) == 0 {
+		log.Infof("Deleting ConfigMap %s since it doesn't contain any cache entries", cm.Name)
```
structured logging

alexec commented on workflow/controller/controller.go:
```diff
@@ -798,6 +806,65 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+			return fmt.Errorf("malformed cache entry: could not unmarshal JSON; unable to parse: %w", err)
+		}
+		if time.Since(entry.LastHitTimestamp.Time) > gcAfterNotHitDuration {
+			log.Infof("Deleting entry with key %s in ConfigMap %s since it's not been hit after %s", key, cm.Name, gcAfterNotHitDuration)
```
structured logging

alexec commented on workflow/controller/controller.go:
```diff
@@ -798,6 +806,65 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
+					continue
+				}
+				if err := wfc.cleanupUnusedCache(ctx, cm, gcAfterNotHitDuration); err != nil {
+					log.WithFields(log.Fields{"key": cm.Name, "error": err}).Errorf("Unable to sync ConfigMap: %s", cm.Name)
```
structured logging

terrytangyuan commented:
I think I've addressed all the feedback besides the one where I left a comment https://github.com/argoproj/argo-workflows/pull/6850#discussion_r751615070. 

@alexec Mind taking another look? Thanks!

alexec commented on docs/environment-variables.md:
```diff
@@ -13,6 +13,8 @@ Note that these environment variables may be removed at any time.
 | `ARCHIVED_WORKFLOW_GC_PERIOD` | `time.Duration` | `24h` | The periodicity for GC of archived workflows. |
 | `ARGO_PPROF` | `bool` | `false` | Enable pprof endpoints |
 | `ARGO_TRACE` | `string` | `"1"` | Whether to enable tracing statements in Argo components. |
+| `CACHE_GC_PERIOD` | `time.Duration` | `"false"` | How often to perform memoization cache GC, which is disabled by default and can be enabled by providing a duration. |
```
maybe add example duration? what about `0s` to disable, rather than "false"?

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,72 @@
+// SyncAllCacheForGC syncs all cache for GC
+func SyncAllCacheForGC(ctx context.Context, configMapInformer cache.SharedIndexInformer, kubeclientset kubernetes.Interface) {
+	gcAfterNotHitDuration := env.LookupEnvDurationOr("CACHE_GC_AFTER_NOT_HIT_DURATION", 30*time.Second)
+	log.WithField("key", "CACHE_GC").Infoln("Cache GC is enabled. Syncing all cache for GC.")
```
something wrong with this WithField?

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,72 @@
+func SyncAllCacheForGC(ctx context.Context, configMapInformer cache.SharedIndexInformer, kubeclientset kubernetes.Interface) {
+	gcAfterNotHitDuration := env.LookupEnvDurationOr("CACHE_GC_AFTER_NOT_HIT_DURATION", 30*time.Second)
+	log.WithField("key", "CACHE_GC").Infoln("Cache GC is enabled. Syncing all cache for GC.")
+	configMaps := configMapInformer.GetIndexer().List()
```
you can use `GetIndexer().ByIndex()` to only return config maps that have the label, no need to check later on

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+// SyncAllCacheForGC syncs all cache for GC
+func SyncAllCacheForGC(ctx context.Context, configMapInformer cache.SharedIndexInformer, kubeclientset kubernetes.Interface) {
```
what not main this a func of `WorkflowController`?  that way you have all the informers etc to hand with `wfc.*`

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,72 @@
+		}
+	}
+	if len(cm.Data) == 0 {
+		log.WithField("configMap", cm.Name).Infoln("Deleting ConfigMap since it doesn't contain any cache entries")
```
minor - `Infoln`?

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,72 @@
+	if len(cm.Data) == 0 {
+		log.WithField("configMap", cm.Name).Infoln("Deleting ConfigMap since it doesn't contain any cache entries")
+		err := kubeclientset.CoreV1().ConfigMaps(cm.Namespace).Delete(ctx, cm.Name, metav1.DeleteOptions{})
+		if err != nil {
```
should you use `apperr.IgnoreNotFound`?

terrytangyuan replied:
Could you clarify what the use case is? Would you like to ignore any errors that are "NotFound"?

alexec replied:
we typically wrap delete with `IgnoreNotFound` to prevent any problems with race conditions

alexec commented on workflow/controller/indexes/configmap_index.go:
```diff
@@ -4,13 +4,20 @@ import (
-const ConfigMapTypeLabel = "workflows.argoproj.io/configmap-type"
+const (
+	// LabelKeyConfigMapType is the label key for the type of configmap.
+	LabelKeyConfigMapType = "workflows.argoproj.io/configmap-type"
```
please move these const to common.go

alexec commented on workflow/controller/controller.go:
```diff
@@ -320,6 +320,18 @@ func (wfc *WorkflowController) startLeading(ctx context.Context, logCtx *log.Ent
 	}
+	if v, found := os.LookupEnv("CACHE_GC_PERIOD"); found {
+		if v != "false" {
+			cacheGCPeriod, err := time.ParseDuration(v)
```
I'm sure there is some `GetEnvDuration(value, defaultValue` func you can use here

alexec commented on workflow/controller/controller.go:
```diff
@@ -320,6 +320,18 @@ func (wfc *WorkflowController) startLeading(ctx context.Context, logCtx *log.Ent
+				log.WithField("CACHE_GC_PERIOD", v).WithError(err).Panic("failed to parse")
+				return
+			}
+			go wait.JitterUntilWithContext(ctx, func(ctx context.Context) {
```
good

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,70 @@
+		}
+	} else {
+		if modified {
+			log.WithField("configMap", cm.Name).Debugln("Updated ConfigMap")
```
I don't think we need this debug logging, `Delete configmaps xxx` should always be logged by the interceptor

alexec commented on workflow/common/common.go:
```diff
@@ -173,6 +173,13 @@ const (
+	// LabelKeyConfigMapType is the label key for the type of configmap.
+	LabelKeyConfigMapType = "workflows.argoproj.io/configmap-type"
+	// LabelValueCacheTypeConfigMap is a key for configmaps that are memoization cache.
+	LabelValueCacheTypeConfigMap = "Cache"
```
suffix? `LabelValueTypeConfigMapCache`

alexec commented on workflow/common/common.go:
```diff
@@ -173,6 +173,13 @@ const (
+	// LabelValueCacheTypeConfigMap is a key for configmaps that are memoization cache.
+	LabelValueCacheTypeConfigMap = "Cache"
+	// LabelValueParameterTypeConfigMap is a key for configmaps that contains parameter values.
+	LabelValueParameterTypeConfigMap = "Parameter"
```
`LabelValueTypeConfigMapParameter` ??

alexec reviewed (commented):
minor suggestions

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,73 @@
+			continue
+		}
+		if err := cleanupUnusedCache(ctx, wfc.kubeclientset, cm, gcAfterNotHitDuration); err != nil {
+			log.WithFields(log.Fields{"configMap": cm.Name, "error": err}).Errorln("Unable to sync ConfigMap")
```
for error, use `WithError`

alexec replied:
use `Error()` not `Errorln()`

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,73 @@
+	gcAfterNotHitDuration := env.LookupEnvDurationOr("CACHE_GC_AFTER_NOT_HIT_DURATION", 30*time.Second)
+	configMaps, err := wfc.configMapInformer.GetIndexer().ByIndex(indexes.ConfigMapLabelsIndex, common.LabelValueTypeConfigMapCache)
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Errorln("Failed to get configmaps from informer")
```
use `Error()` not `Errorln()`

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,73 @@
+			return fmt.Errorf("malformed cache entry: could not unmarshal JSON; unable to parse: %w", err)
+		}
+		if time.Since(entry.LastHitTimestamp.Time) > gcAfterNotHitDuration {
+			log.WithFields(log.Fields{"key": key, "configMap": cm.Name, "gcAfterNotHitDuration": gcAfterNotHitDuration}).Infoln("Deleting entry in ConfigMap since it's not been hit")
```
use `Info()` not `Infoln()`

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+// SyncAllCacheForGC syncs all cache for GC
+func (wfc *WorkflowController) SyncAllCacheForGC(ctx context.Context) {
```
minor - can be private

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,73 @@
+
+// SyncAllCacheForGC syncs all cache for GC
+func (wfc *WorkflowController) SyncAllCacheForGC(ctx context.Context) {
+	gcAfterNotHitDuration := env.LookupEnvDurationOr("CACHE_GC_AFTER_NOT_HIT_DURATION", 30*time.Second)
```
could we 

* make global 
* print value on start-up, e.g. using `init()`

terrytangyuan replied:
Good idea. Done

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -18,22 +18,27 @@ import (
+var gcAfterNotHitDuration = env.LookupEnvDurationOr("CACHE_GC_AFTER_NOT_HIT_DURATION", 30*time.Second)
+
+func init() {
+	log.Infof("CACHE_GC_AFTER_NOT_HIT_DURATION: %s", gcAfterNotHitDuration)
```
please use structured logging

terrytangyuan replied:
Fixed

alexec commented on workflow/controller/controller.go:
```diff
@@ -320,6 +320,12 @@ func (wfc *WorkflowController) startLeading(ctx context.Context, logCtx *log.Ent
 	for i := 0; i < podWorkers; i++ {
 		go wait.Until(wfc.podWorker, time.Second, ctx.Done())
 	}
+	cacheGCPeriod := env.LookupEnvDurationOr("CACHE_GC_PERIOD", 0)
```
global and log in init() too?

alexec replied:
should say, it is useful for users to be able to check their config

terrytangyuan replied:
Oh I missed this comment. Just updated. It might be a good idea to log all env vars at the start of the controller (in a potential follow-up PR).

alexec commented on workflow/controller/controller.go:
```diff
@@ -320,6 +320,12 @@ func (wfc *WorkflowController) startLeading(ctx context.Context, logCtx *log.Ent
 	}
+	cacheGCPeriod := env.LookupEnvDurationOr("CACHE_GC_PERIOD", 0)
+	if cacheGCPeriod != 0 {
+		go wait.JitterUntilWithContext(ctx, func(ctx context.Context) {
```
minor - but I think this could be:

```golang
		go wait.JitterUntilWithContext(ctx, wfc.syncAllCacheForGC, cacheGCPeriod, 0.0, true)
```

terrytangyuan replied:
Thanks! Updated.

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,78 @@
+		}
+		if err := cleanupUnusedCache(ctx, wfc.kubeclientset, cm, gcAfterNotHitDuration); err != nil {
+			log.WithField("configMap", cm.Name).WithError(err).Error("Unable to sync ConfigMap")
+			continue
```
minor - `continue` not needed

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,78 @@
+func (wfc *WorkflowController) syncAllCacheForGC(ctx context.Context) {
+	configMaps, err := wfc.configMapInformer.GetIndexer().ByIndex(indexes.ConfigMapLabelsIndex, common.LabelValueTypeConfigMapCache)
+	if err != nil {
+		log.WithError(err).Error("Failed to get configmaps from informer")
```
minor - probably should return here

alexec commented on workflow/controller/cache_gc.go:
```diff
@@ -0,0 +1,78 @@
+			log.WithField("configMap", cm.Name).Error("Unable to convert object to configmap when syncing ConfigMaps")
+			continue
+		}
+		if err := cleanupUnusedCache(ctx, wfc.kubeclientset, cm, gcAfterNotHitDuration); err != nil {
```
* no need for global as parameter here?
* if `cleanupUnusedCache` was `wfc.cleanupUnusedCache` then it could access `wfc.kubeclientset`

alexec reviewed (commented):
some suggestions

terrytangyuan commented:
Addressed additional comments.
